### DISCUSSION TITLE
form data validation using wtf forms

### DISCUSSION THREAD

**Discussion by sandeep-gh** [elimintz/justpy#575] [General]
I was attempting to use wtforms (https://flask-wtf.readthedocs.io/en/1.0.x/) to bake in data validation. Turned out to be much harder than I had envisioned. 
First, the event data used in wtforms in some sort of multidict. It seems that the request json object could have multiple entries for same keys which necessitates use of multidict. 
If so, we should look into implications for justpy. 

Second, and the main reason why this didn't work out, is because I am not able to associate/link components at creation time with 
data derived from event handling. 
So, when form is created its component looks as follows:
 
```
('name', Input(id: name, html_tag: input, input_type: text, vue_type: html_component, value: , checked: False, number of components: 0)), ('email', Input(id: email, html_tag: input, input_type: text, vue_type: html_component, value: , checked: False, number of components: 0)
```

but the data in event handler is :
```
on_submit click =  [{'html_tag': 'input', 'id': '11', 'type': 'text', 'value': 'sdsf', 'checked': False}, 
{'html_tag': 'input', 'id': '12', 'type': 'text', 'value': 'sdsd', 'checked': False}, ...}]
```
The id's are different for the form components. Here is the unfinished code if case anyone has ideas or want to take a stab at it: 
https://github.com/sandeep-gh/justpy_form_wtf_validation/blob/main/jp_and_wtforms.py

**Comment by WolfgangFahl** [comment]
@sandeep-gh  - very interesting analysis. Actually Tim and I come from a flask background e.g. we worked on 
https://github.com/WolfgangFahl/pyFlaskBootstrap4 with a demo at
http://fb4demo.bitplan.com/
before justpy got interesting for us since it's much more reactive.
For Flask WFT Forms we created our own set of extra widgets see https://github.com/WolfgangFahl/pyFlaskBootstrap4/blob/main/fb4/widgets.py which were sort of an inspiration for https://github.com/WolfgangFahl/pyJustpyWidgets

Integrating all these different frameworks is unfortunately much harder then we would wish since core softare architecture and software engineering aspects have been having  a low priority in most of the frameworks involved and you'll see ugly workarounds and APIs all over. Our own code is unfortunately not exempt - we also chimed in in doing work-around more often than proper solutions. So it would be good if we would find better integration ways to make the transition from one framework to another and the interoperability between the different frameworks much smoother.

**Reply by sandeep-gh** [reply]
For FlaskWTF, I was only interested in their data validation feature. The rendering engine for FlaskWTF is anyways cumbersome to be used from justpy. Even using data validation turned out to be hard  because FlaskWTF decided to go with a custom multidict object.

**Comment by sandeep-gh** [comment]
The other reason is the ID generation aspect of justpy. This needs a more careful look as to why we can't stick with same id when json is returned from frontend. For my project, I might not use justpy id generation but instead assign them externally.
